File: 67-add-binary/add-binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        i = len(a) - 1
        j = len(b) - 1
        carry = 0
        result = []

        while i >= 0 or j >= 0 or carry:
            total = carry
            if i >= 0:
                total += int(a[i])
                i -= 1
            if j >= 0:
                total += int(b[j])
                j -= 1   

            # Append the binary digit
            result.append(str(total % 2))

            # Update Carry
            carry = total // 2
             
        # Reverse because we built it backwards
        return "".join(reversed(result))         

File: 67-add-binary/add-binary.java
class Solution {
    public String addBinary(String a, String b) {
        int i = a.length() - 1;
      int j = b.length() - 1;
      int carry = 0;
      
      StringBuilder result = new StringBuilder();
      
      while (i >= 0 || j >= 0 || carry == 1) {
         int bitA = (i >= 0) ? a.charAt(i) - '0' : 0;
         int bitB = (j >= 0) ? b.charAt(j) - '0' : 0;
         
         int sum = bitA + bitB + carry;
         
         result.append(sum % 2);
         carry = sum / 2;
         
         i--;
         j--;
      } 
      return result.reverse().toString();
    }
}